kernel/scheduler_priority/src/lib.rs

```diff
@@ -4,43 +4,157 @@
-pub fn select_next_task(apic_id: u8) -> Option<TaskRef> {
-    let mut runqueue_locked = match RunQueue::get_runqueue(apic_id) {
-        Some(rq) => rq.write(),
-        _ => {
-            error!("BUG: select_next_task_priority(): couldn't get runqueue for core {apic_id}",);
-            return None;
+impl Scheduler {
+    pub fn new(idle_task: TaskRef) -> Self {
+        Self {
+            idle_task,
+            queue: BinaryHeap::new(),
         }
-    };
-
-    // This is a temporary solution before the PR to only store runnable tasks in
-    // the run queue is merged.
-    let mut blocked_tasks = Vec::with_capacity(2);
-    while let Some(mut task) = runqueue_locked.pop() {
-        if task.is_runnable() {
-            for t in blocked_tasks {
-                runqueue_locked.push(t)
+    }
+}
+
+impl task::scheduler::Scheduler for Scheduler {
+    fn next(&mut self) -> TaskRef {
+        // This is a temporary solution before the PR to only store runnable tasks in
+        // the run queue is merged.
+        let mut blocked_tasks = Vec::with_capacity(2);
+        while let Some(mut task) = self.queue.pop() {
+            if task.task.is_runnable() {
+                for t in blocked_tasks {
+                    self.queue.push(t)
+                }
+                task.last_ran = time::now::<time::Monotonic>();
+                self.queue.push(task.clone());
+                return task.task;
+            } else {
+                blocked_tasks.push(task);
             }
-            task.last_ran = time::now::<time::Monotonic>();
-            runqueue_locked.push(task.clone());
-            return Some(task.task);
+        }
+        for task in blocked_tasks {
+            self.queue.push(task);
+        }
+        self.idle_task.clone()
+    }
+
+    fn add(&mut self, task: TaskRef) {
+        self.queue
+            .push(PriorityTaskRef::new(task, DEFAULT_PRIORITY));
+    }
+
+    fn busyness(&self) -> usize {
+        self.queue.len()
+    }
+
+    fn remove(&mut self, task: &TaskRef) -> bool {
+        let old_len = self.queue.len();
+        self.queue
+            .retain(|priority_task| priority_task.task != *task);
+        let new_len = self.queue.len();
+        // We should have removed at most one task from the run queue.
+        debug_assert!(
+            old_len - new_len < 2,
+            "difference between run queue lengths was: {}",
+            old_len - new_len
+        );
+        new_len != old_len
+    }
+
+    fn as_priority_scheduler(&mut self) -> Option<&mut dyn task::scheduler::PriorityScheduler> {
+        Some(self)
+    }
+
+    fn drain(&mut self) -> alloc::boxed::Box<dyn Iterator<Item = TaskRef> + '_> {
+        Box::new(self.queue.drain().map(|priority_task| priority_task.task))
+    }
+
+    fn dump(&self) -> Vec<TaskRef> {
+        self.queue
+            .clone()
+            .into_iter()
+            .map(|priority_task| priority_task.task)
+            .collect()
```

Comment:
We can't return an associated type because then the trait wouldn't be object safe.

> Could also re-name it more accurately to something like task_iter, or a similar name of your choosing.

As I've stated above it can't return an iterator so I've renamed it to `tasks` because it returns a vec of the tasks in the run queue.

Reply:
Right, I didn't mean to suggest using an associated type like my old scheduler refactor design, but rather an `impl Iterator<...>`. But I just realized that I had only recently heard about [RPITIT](https://github.com/rust-lang/rust/issues/91611) being supported in soon-to-come future versions of Rust, so I was mixing up the fact that RPITIT support is coming soon with the fact that it doesn't actually exist yet in the version of Rust we're pinned to.